File: integration-test/src/test/js/steps-filters.js
const { strictEqual } = require('assert');
const { After, Given, Then, When } = require('@cucumber/cucumber');
const { Gene, GeneID, TokenTypeFilter } = require('wallet-lib');
const {
    clientCreatesIdForTokenOfType,
    clientIssuesTokenForClientWithValue,
    clientRegistersTokenType,
    thereIsAClient,
    tokenTypesContextMixin,
    tokensContextMixin,
    walletContextMixin,
} = require('./cnft-common-steps');
const { Ascii, createPublicApi, log } = require('./cnft-util');

const { FgGreen: FgBase, FgYellow: FgEvt, FgMagenta: FgVal, Reset } = Ascii;

const context = {};
Object.assign(context, walletContextMixin, tokenTypesContextMixin, tokensContextMixin);

Given(
    '[Filters] There is a client {string}',
    async function (clientName) {
        return await thereIsAClient(context, clientName);
    }
);

When(
    '[Filters] Client {string} registers token type {string} with fungible gene',
    async function (clientName, tokenTypeName) {
        log(`${FgBase}Client ${FgVal}'${clientName}'${FgBase} registers token type ${FgVal}'${tokenTypeName}'` +
            `${FgBase} with fungible gene${Reset}`);
        await clientRegistersTokenType(context, clientName, tokenTypeName, Gene(GeneID.Fungible, []));
    }
);

When(
    '[Filters] Client {string} registers token type {string} without fungible gene',
    async function (clientName, tokenTypeName) {
        log(`${FgBase}Client ${FgVal}'${clientName}'${FgBase} registers token type ${FgVal}'${tokenTypeName}'` +
            `${FgBase} without fungible gene${Reset}`);
        await clientRegistersTokenType(context, clientName, tokenTypeName);
    }
);

When(
    '[Filters] Client {string} creates ID for token {string} of type {string}',
    async (clientName, tokenName, tokenTypeName) =>
        await clientCreatesIdForTokenOfType(context, clientName, tokenName, tokenTypeName)
);

When(
    '[Filters] Client {string} issues token {string} for client {string} with value {string}',
    async function (issuerName, tokenName, clientName, tokenValue) {
        log(`${FgBase}Client ${FgVal}'${issuerName}'${FgBase} issues token ${FgVal}'${tokenName}'${FgBase}` +
            ` for client ${FgVal}'${clientName}'${FgBase} with value ${FgVal}'${tokenValue}'${Reset}`);

        const clientWallet = context.getWalletByClientName(clientName);
        const tokenIssued = clientWallet.waiterFor(
            'onWalletEvents',
            (wallet, { owner }) => {
                log(`${FgEvt}Got ${FgVal}'onWalletEvents'${FgEvt}:${Reset}`);
                if (owner.tokensReceived.length) {
                    log(`${FgEvt}  - tokensReceived: ${FgVal}${owner.tokensReceived}${Reset}`);
                    return true;
                }
                return false;
            }
        );

        await clientIssuesTokenForClientWithValue(context, issuerName, clientName, tokenName, tokenValue);

        await tokenIssued;
        log(`${FgBase}  Done: issued token ${FgVal}'${tokenName}'${FgBase} for ${FgVal}'${clientName}'${Reset}\n`);
    }
);

async function clientHaveTokenTypesFiltered(clientName, tokenTypesCountExpected, changeGeneNegation) {
    const publicApi = createPublicApi();
    const walletIdentity = context.getWalletIdentityByClientName(clientName);
    const tokenTypesFiltered = await publicApi.listTokenTypesFiltered(
        TokenTypeFilter(
            GeneID.Fungible,
            changeGeneNegation
        )
    );
    const myTokenTypes = tokenTypesFiltered.filter(tt => tt.issuerId === walletIdentity);
    strictEqual(myTokenTypes.length, tokenTypesCountExpected);
}

Then(
    '[Filters] Client {string} have {int} token types filtered by fungible',
    async function (clientName, expectedTokenTypesCount) {
        log(`${FgBase}There are ${FgVal}${expectedTokenTypesCount}${FgBase} token types filtered by fungible${Reset}`);
        await clientHaveTokenTypesFiltered(clientName, expectedTokenTypesCount, false);
        log(`${FgBase}  Done: correct!${Reset}\n`);
    }
);

Then(
    '[Filters] Client {string} have {int} token types filtered by non-fungible',
    async function (clientName, expectedTokenTypesCount) {
        log(`${FgBase}There are ${FgVal}${expectedTokenTypesCount}${FgBase} token types filtered by non-fungible${Reset}`);
        await clientHaveTokenTypesFiltered(clientName, expectedTokenTypesCount, true);
        log(`${FgBase}  Done: correct!${Reset}\n`);
    }
);

Then(
    '[Filters] Client {string} sees {int} token\\(s) in his list filtered by fungible',
    async function (clientName, expectedTokensCount) {
        log(`${FgBase}Client ${FgVal}'${clientName}'${FgBase} sees ${FgVal}${expectedTokensCount}${FgBase} ` +
            `token(s) in his list filtered by fungible${Reset}`);

        const wallet = context.getWalletByClientName(clientName);
        const tokensFiltered = await wallet.listTokensFiltered(TokenTypeFilter(GeneID.Fungible, false));
        strictEqual(tokensFiltered.length, expectedTokensCount);

        log(`${FgBase}  Done: correct!${Reset}\n`);
    }
);

Then(
    '[Filters] Client {string} sees {int} token\\(s) in his list filtered by non-fungible',
    async function (clientName, expectedTokensCount) {
        log(`${FgBase}Client ${FgVal}'${clientName}'${FgBase} sees ${FgVal}${expectedTokensCount}${FgBase} ` +
            `token(s) in his list filtered by non-fungible${Reset}`);

        const wallet = context.getWalletByClientName(clientName);
        const tokensFiltered = await wallet.listTokensFiltered(TokenTypeFilter(GeneID.Fungible, true));
        strictEqual(tokensFiltered.length, expectedTokensCount);

        log(`${FgBase}  Done: correct!${Reset}\n`);
    }
);

After(function () {
/*    return new Promise((resolve, _reject) => {
        log(`Stopping wallets ...`);
        for (const index in context.wallets) {
            log(`Stopping wallet ${index} ...`);
        }
        log(`Wallets stopped.`);
        return resolve();
    });*/
});

File: integration-test/src/test/scala/ru/sberbank/blockchain/cnft/wallet/test/StepsFilters.scala
package ru.sberbank.blockchain.cnft.wallet.test

import io.cucumber.scala.{EN, ScalaDsl}
import org.slf4j.LoggerFactory
import ru.sberbank.blockchain.cnft.common.types._
import ru.sberbank.blockchain.cnft.model._
import ru.sberbank.blockchain.cnft.wallet.dsl._
import ru.sberbank.blockchain.cnft.wallet.test.CNFTCommonSteps._
import ru.sberbank.blockchain.cnft.wallet.test.CNFTUtil.Chain
import ru.sberbank.blockchain.cnft.wallet.test.Util.{mkTokenTypesString, mkTokensString}

class StepsFilters extends ScalaDsl with EN {

    private val log = LoggerFactory.getLogger(classOf[StepsFilters])

    //    After { _: Scenario =>
    //        log.info("Stopping wallets ...")
    //        context.wallets.values.foreach(_.stopListenBlocks())
    //        log.info("All wallets stopped.")
    //    }

    Given("""[Filters] There is a client {string}""") {
        clientName: String =>
            log.info(s"Creating new wallet for $clientName")
            thereIsAClient(context, clientName)
            log.info(s"Created new client $clientName")
    }

    When("""[Filters] Client {string} registers token type {string} with fungible gene""") {
        (clientName: String, tokenTypeName: String) =>
            log.info(s"Registering fungible token type $tokenTypeName for $clientName...")
            clientRegistersTokenType(
                context,
                clientName,
                tokenTypeName,
                Some(
                    Gene(
                        id = GeneID.Fungible,
                        parameters = Collection.empty
                    )
                )
            )
            context.withTokenType(tokenTypeName) { tokenTypeId =>
                log.info(s"Registered fungible token type $tokenTypeName for $clientName ($tokenTypeId)")
            }
    }

    When("""[Filters] Client {string} registers token type {string} without fungible gene""") {
        (clientName: String, tokenTypeName: String) =>
            log.info(s"Registering non-fungible token type $tokenTypeName for $clientName...")
            clientRegistersTokenType(context, clientName, tokenTypeName)
            context.withTokenType(tokenTypeName) { tokenTypeId =>
                log.info(s"Registered non-fungible token type $tokenTypeName for $clientName ($tokenTypeId)")
            }
    }

    When("""[Filters] Client {string} creates ID for token {string} of type {string}""") {
        (clientName: String, tokenName: String, tokenTypeName: String) =>
            clientCreatesIdForTokenOfType(context, clientName, tokenName, tokenTypeName)
    }

    When("""[Filters] Client {string} issues token {string} for client {string} with value {string}""") {
        (issuerName: String, tokenName: String, clientName: String, tokenValue: String) =>
            log.info(s"Issuing token '$tokenName' for $clientName ...")

            val clientWallet = context.wallets.getOrElse(clientName, throw new Exception(s"Unknown client $clientName"))
            val issueTokenEvent = Util.awaitWalletEvents(clientWallet) { case (_, events) =>
                val tokensReceived = events.owner.tokensReceived
                if (tokensReceived.nonEmpty)
                    log.info(s"[ClientFilter Listener] New token received: ${tokensReceived.toSeq}.")

                tokensReceived.headOption
            }

            clientIssuesTokenForClientWithValue(
                context,
                issuerName,
                clientName,
                tokenName,
                tokenValue
            )
            issueTokenEvent.get
            log.info(s"Issued token '$tokenName' for '$clientName'")
    }

    private def clientHaveTokenTypesFiltered(
        clientName: String,
        tokenTypesCountExpected: Int,
        changeGeneNegation: Boolean
    ): Result[Unit] = {
        for {
            walletIdentity <- context.getWalletIdentityByClientName(clientName)
            tokenTypesFiltered <-
                Chain
                    .listTokenTypesFiltered(
                        TokenTypeFilter(
                            changeGeneId = GeneID.Fungible,
                            negation = changeGeneNegation
                        )
                    )
            myTokenTypes = tokenTypesFiltered.filter(_.issuerId == walletIdentity)
        } yield {
            log.info(mkTokenTypesString(myTokenTypes))
            assert(myTokenTypes.length == tokenTypesCountExpected)
        }
    }

    Then("""[Filters] Client {string} have {int} token types filtered by fungible""") {
        (clientName: String, tokenTypesCountExpected: Int) =>
            log.info("Checks fungible token types")
            clientHaveTokenTypesFiltered(clientName, tokenTypesCountExpected, changeGeneNegation = false)
                .orFail("Failed to check fungible token types")
    }

    Then("""[Filters] Client {string} have {int} token types filtered by non-fungible""") {
        (clientName: String, tokenTypesCountExpected: Int) =>
            log.info("Checks non-fungible token types")
            clientHaveTokenTypesFiltered(clientName, tokenTypesCountExpected, changeGeneNegation = true)
                .orFail("Failed to check non-fungible token types")
    }

    Then("""[Filters] Client {string} sees {int} token\(s) in his list filtered by fungible""") {
        (clientName: String, expectedTokensCount: Int) =>
            log.info(s"Checks fungible tokens for $clientName...")
            context.withWallet(clientName) { wallet =>
                val tokensFiltered =
                    wallet
                        .listTokensFiltered(
                            TokenTypeFilter(
                                changeGeneId = GeneID.Fungible,
                                negation = false
                            )
                        )
                        .orFail(s"Failed to check fungible tokens for $clientName")
                log.info(mkTokensString(tokensFiltered))
                assert(tokensFiltered.length == expectedTokensCount)
            }
    }

    Then("""[Filters] Client {string} sees {int} token\(s) in his list filtered by non-fungible""") {
        (clientName: String, expectedTokensCount: Int) =>
            log.info(s"Checks non-fungible tokens for $clientName...")
            context.withWallet(clientName) { wallet =>
                val tokensFiltered =
                    wallet
                        .listTokensFiltered(
                            TokenTypeFilter(
                                changeGeneId = GeneID.Fungible,
                                negation = true
                            )
                        )
                        .orFail(s"Failed to check non-fungible tokens for $clientName")
                log.info(mkTokensString(tokensFiltered))
                assert(tokensFiltered.length == expectedTokensCount)
            }

    }

    private val _context = new ThreadLocal[ScenarioContext]() {
        override def initialValue(): ScenarioContext = new ScenarioContext
    }

    private def context: ScenarioContext = _context.get()

    class ScenarioContext extends WalletsContext with TokenTypesContext with TokensContext

}
